test/e2e/k8s_util.go

```diff
@@ -0,0 +1,453 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	v1net "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+
+	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
+	secClient "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned/typed/security/v1alpha1"
+)
+
+type KubernetesUtils struct {
+	*TestData
+	podCache       map[string][]v1.Pod
+	ClientSet      *kubernetes.Clientset
+	SecurityClient *secClient.SecurityV1alpha1Client
+}
+
+func NewKubernetesUtils(data *TestData) (*KubernetesUtils, error) {
+	return &KubernetesUtils{
+		TestData:       data,
+		podCache:       map[string][]v1.Pod{},
+		ClientSet:      data.clientset.(*kubernetes.Clientset),
+		SecurityClient: data.securityClient.(*secClient.SecurityV1alpha1Client),
+	}, nil
+}
+
+// GetPod returns a pod with the matching namespace and name
+func (k *KubernetesUtils) GetPod(ns string, name string) (*v1.Pod, error) {
+	pods, err := k.getPodsUncached(ns, "pod", name)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "unable to get pod %s/%s", ns, name)
+	}
+	if len(pods) == 0 {
+		return nil, nil
+	}
+	return &pods[0], nil
+}
+
+func (k *KubernetesUtils) getPodsUncached(ns string, key, val string) ([]v1.Pod, error) {
+	v1PodList, err := k.ClientSet.CoreV1().Pods(ns).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%v=%v", key, val),
+	})
+	if err != nil {
+		return nil, errors.WithMessage(err, "unable to list pods")
+	}
+	return v1PodList.Items, nil
+}
+
+// GetPods returns an array of all pods in the given namespace having a k/v label pair.
+func (k *KubernetesUtils) GetPods(ns string, key string, val string) ([]v1.Pod, error) {
+	if p, ok := k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)]; ok {
+		return p, nil
+	}
+
+	v1PodList, err := k.getPodsUncached(ns, key, val)
+	if err != nil {
+		return nil, errors.WithMessage(err, "unable to list pods")
+	}
+	k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)] = v1PodList
+	return v1PodList, nil
+}
+
+// Probe execs into a pod and checks its connectivity to another pod.  Of course it assumes
+// that the target pod is serving on the input port, and also that wget is installed.  For perf it uses
+// spider rather then actually getting the full contents.
```

Comment:
"For perf it used spider rather then actually getting the full contents."

This comment no longer applied. It dates back to the use of "wget", but then I switched it to "ncat". It's my bad but would appreciate it if you could update the comment :)

Reply:
Done